Step3 - 로또(자동)

piopoi opened the pull request:
안녕하세요. 리뷰어님.
3단계 PR드립니다.

개인적으로 이번 3단계 미션으로 얻은게 너무 많습니다.
일급 콜렉션, 원시값 포장, 정적 팩토리 메서드 등등 너무 많은 개념을 처음부터 적용시키려다 보니 설계가 산으로 가고 머리속은 꼬여가서 제출하는데 시간이 너무 오래 걸렸습니다.
중간에 2번 정도 대대적으로 리팩토링을 진행하긴 했지만, 토비님 말씀처럼 차라리 개발한 소스를 싹 날리고 다시 시작하는 것도 괜찮지 않았을까 하는 생각이 드네요.
설계와 리펙토링 방식에 대한 중요성을 크게 체감한 미션이었던 것 같습니다.

piopoi committed:
refactor(StringAddCalculator): 2단계 계산기 리뷰내용 반영

piopoi committed:
docs(README.md): 로또(자동) 기능 추가

piopoi committed:
docs(README.md): 구매할 수 있는 로또 개수 반환과 로또 생성 기능 분리

piopoi committed:
test(PayTest): 구매 가능한 로또 개수 반환 테스트 추가

piopoi committed:
feat(Pay): 구매 가능한 로또 개수 반환 기능 추가

piopoi committed:
docs(README.md): 기능 분리

piopoi committed:
test(LottoTest): 로또 생성 테스트 추가

piopoi committed:
feat(Lotto): 로또 생성 기능 추가

piopoi committed:
feat(Lotto): 자동번호 로또 1개 생성 기능 추가

piopoi committed:
test(PayTest): 구매가능개수 확인 테스트 수정

piopoi committed:
docs(README.md): 당첨번호로 로또 생성 기능 수정

piopoi committed:
feat(LottoGame): LottoGame 기능 추가

LottoGame 기능 추가
- Pay에 있던 구입가능개수 반환기능 이동
- 구입가능개수만큼 로또 생성하는 기능 추가
- 지난주 당첨번호를 입력받아 로또 생성하는 기능 추가
- 위 3가지 기능에 대한 테스트 추가

piopoi committed:
refactor: Lotto 클래스 분리

Lotto 클래스 분리
- LottoNumber : 원시값 포장
- LottoFactory : 정적 팩토리 메서드 사용
- 관련 테스트 수정

piopoi committed:
docs(README.md): 기능 목록 수정

piopoi committed:
refactor: LottoFactory, LottoNumber 관련 수정

1. LottoFactory 추가
- 자동/수동 로또 생성
- 정적 팩토리 메서드 적용
2. LottoNumber
- 캐시 추가
3. LottoFactory, LottoNumber 수정과 관련된 프로덕션/테스트 코드 수정

piopoi committed:
feat(InputView): 금액, 당첨번호 입력 기능 추가

piopoi committed:
feat: 로또 출력을 위한 수정

로또번호 정렬 후 출력을 위한 수정
- Lotto, LottoNumber : toString 수정
- LottoNumber 정렬을 위해 Comparable 인터페이스 구현

piopoi committed:
docs(README.md): 로또 문서 분리

기존 step1~3의 기능사항이 함께 정의되었던 문서에서 로또 관련 내용을 분리.

piopoi committed:
feat(step3): 패키지 구조 변경

domain, view 패키지로 분리

piopoi committed:
docs(README.md): 당첨금 정보 추가

piopoi committed:
docs(README.md): 당첨결과 반영 기능 추가

piopoi committed:
docs(README.md): 기능 목록 수정

piopoi committed:
refactor(step3.domain): 로또 생성 및 당첨번호 생성 로직 변경

1. 로또 생성자 파라미터 변경
- 원시값으로 로또번호를 받아서 생성자에서 LottoNumber로 Parsing.
2. 당첨번호는 LottoNumber로 Parsing 안하도록 수정
3. 관련 프로덕션/테스트 소스 수정

piopoi committed:
feat(Ranking): 로또 당첨순위 Enum, test 추가

piopoi committed:
feat(LottoGame): 로또 게임 실행 기능/테스트 추가.

piopoi committed:
feat(LottoResult): 로또 당첨결과 기능/테스트 추가

1. LottoResult
- 로또 건 별 당첨순위를 받아서 지속적으로 UPDATE.
- 수익률 계산
- 특정 순위 당첨횟수 조회
- 게임결과 출력: 순위 별 당첨횟수, 수익률
2. LottoResultTest
- 순위별 당첨횟수 확인 테스트
- 로또 별 당첨결과 반영 테스트
- 수익률 계산 테스트

piopoi committed:
feat(Lottos): 구입한 로또 클래스/테스트 추가

1. Lottos
- 구입한 모든 로또에 대한 당첨결과 반환
- 모든 로또 번호 출력
2. LottosTest
- 구입한 모든 로또와 당첨번호 매칭 테스트

piopoi committed:
feat(Lotto): 로또 1개의 당첨결과 확인 기능 추가

1. Lotto
- 로또 1개의 당첨결과 확인 기능
2. LottoTest
- 로또 1개의 당첨결과 확인 기능 테스트

piopoi committed:
feat(OutputView): 출력 기능 추가

출력 기능 추가
- 구입한 로또의 로또번호 전체 출력
- 당첨 통계 출력

piopoi committed:
feat(LottoMain): 로또 게임 실행을 위한 Main 클래스 추가

piopoi committed:
docs(README.md): step1, step2 문서 분리

parkeeseul commented on src/main/java/step3/domain/Lotto.java:
```diff
@@ -0,0 +1,66 @@
+import java.util.stream.Collectors;
+
+public class Lotto {
+    private static final int LOTTO_SIZE = 6;
```
`Lotto.LOTTO_SIZE` VS `Lotto.SIZE`  어떻게 생각하시나요? 😄

piopoi replied:
중복을 없애니 더 인식하기가 쉬워지네요.
수정하겠습니다. 😃

parkeeseul commented on src/main/java/step3/domain/Lotto.java:
```diff
@@ -0,0 +1,66 @@
+
+    private final Set<LottoNumber> lotto;
+
+    Lotto(List<Integer> numbers) {
+        this.lotto = numbers.stream()
+                .map(LottoNumber::of)
+                .collect(Collectors.toSet());
+        validate();
+    }
```
검증 후 값을 할당하는 것이 조금 더 자연스러울 것 같은데 어떻게 생각하시나요? 😄

piopoi replied:
인자값으로 받은 numbers를 선 체크 후 세팅하도록 수정하였습니다. 😄

parkeeseul commented on src/main/java/step3/domain/Lotto.java:
```diff
@@ -0,0 +1,66 @@
+
+    private void validate() {
+        if (isInvalidNumberCount()) {
+            throw new IllegalArgumentException("중복되지 않은 6개의 숫자를 입력해주세요.");
```
정의해둔 LOTTO_SIZE 를 사용하면 만약 로또 사이즈가 변경되었을 경우 에러 메세지를 수정해 주지 않아도 될 것 같아요 😄

piopoi replied:
수정하였습니다. 👍

parkeeseul commented on src/main/java/step3/domain/Lotto.java:
```diff
@@ -0,0 +1,66 @@
+        return LOTTO_SIZE;
+    }
+
+    int match(List<Integer> winnerNumbers) {
```
접근 제어자를 생략해 주신 이유가 있으신가요? 😄

piopoi replied:
LottoTest 에서 테스트하기 위해 default로 두었습니다.
토비님께서 테스트를 위해 접근제어자를 default로 수정하던 부분이 기억나서 따라해보았습니다. 😄

parkeeseul replied:
해당 메서드는 외부(Lottos)에서 사용하는 메서드 이기 때문에 테스트 때문에 default 로 선언해 주지 않아도 될 것 같아요!
다만 동일 패키지 내부에서만 사용하도록 하시려는 의도였다면 default 로 두어도 좋을 것 같네요 😄

piopoi replied:
public이라는 선택지를 잊고 있었네요 😢 
수정하겠습니다!!

parkeeseul commented on src/main/java/step3/domain/LottoGame.java:
```diff
@@ -0,0 +1,40 @@
+import step3.view.OutputView;
+
+public class LottoGame {
+    private static final int PRICE_LOTTO = 1000;
```
돈과 같은 단위는 언더바를 통해 가독성을 높여볼 수 있을 것 같아요 😄 
```suggestion
    private static final int PRICE_LOTTO = 1_000;
```

piopoi replied:
int 값에 언더바 추가가 가능하단 것을 처음 알았네요..
감사합니다. 수정하겠습니다!

parkeeseul commented on src/main/java/step3/domain/LottoNumber.java:
```diff
@@ -0,0 +1,65 @@
+        return lottoNumberCache.get(number);
+    }
+
+    public static int getMaxLottoNumber() {
+        return MAX_LOTTO_NUMBER;
+    }
+
+    public static int getMinLottoNumber() {
+        return MIN_LOTTO_NUMBER;
+    }
```
해당 상수들이 필요하다면 접근제어자를 변경하여 외부에서도 사용할 수 있도록 해보면 어떨까요? 😄

piopoi replied:
내부 변수들의 경우에는 직접 접근하지 말고 메시지를 보내라고 하셨었는데,
혹시 상수의 경우는 외부에 공개하는 방식으로 많이 사용되는 부분일까요?
그 동안 상수/변수 구분 없이 습관처럼 get 메서드를 써왔던 것 같아 질문드립니다!
소스는 수정하겠습니다 😄

piopoi replied:
내부 변수들의 경우에는 직접 접근하지 말고 메시지를 보내라고 하셨었는데,
혹시 상수의 경우는 외부에 공개하는 방식으로 많이 사용되는 부분일까요?
그 동안 상수/변수 구분 없이 습관처럼 get 메서드를 써왔던 것 같아 질문드립니다!
소스는 수정하겠습니다 😄

parkeeseul replied:
상수와 변수는 조금 다를 것 같아요.
객체의 내부의 인스턴스 변수들은 객체의 상태를 나타내고 객체가 가지는 행위들에서 사용되기 때문에
행위에 따라 변경이 되기도 하는데요, 외부에서 상태값에 직접 접근하여 변경하거나 getter 를 사용한다면 비즈니스 로직이 여러곳에 분산될 위험이 있있을 것 같아요.
반면에 상수는 행위에 따라 변경되지 않고 고정된 값으로 존재하는데요, 때문에 외부에서 참조해도 무방하다고 생각합니다. 😄 
어떻게 생각하시나요?

parkeeseul commented on src/main/java/step3/domain/LottoResult.java:
```diff
@@ -0,0 +1,46 @@
+        return rankingMap.get(ranking);
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder text = new StringBuilder();
+        List<Ranking> rankings = Arrays.stream(Ranking.values())
+                .filter(ranking -> ranking != Ranking.NONE)
+                .collect(Collectors.toList());
+        for (Ranking ranking : rankings) {
+            text.append(String.format("%s- %d개\n", ranking.toString(), rankingCount(ranking)));
+        }
+        text.append(String.format("총 수익률은 %.02f입니다.", yield));
+        return text.toString();
+    }
```
핵심로직을 담당하는 도메인에 UI 로직이 혼재되어 있네요
분리해 보면 어떨까요? 😄 
(다른 부분들도 함깨 확인 부탁드려요 )
```
프로그래밍 요구사항
- 핵심 로직을 구현하는 코드와 UI를 담당하는 로직을 구분한다.
```

piopoi replied:
중요한 요구사항을 놓쳤네요 😢 
수정하여 커밋하겠습니다.

parkeeseul commented on src/main/java/step3/domain/Lottos.java:
```diff
@@ -0,0 +1,28 @@
+        this.lottos = lottos;
+    }
+
+    public LottoResult allMatch(List<Integer> winnerNumbers) {
```
우승 번호가 6개를 초과하여 전달되는 경우가 있다면 어떨게 될까요? 🤔

piopoi replied:
WinnerLotto를 따로 생성하여 내부에 Lotto 객체를 가지도록 수정하였습니다.
로또 번호 개수 체크는 Lotto 객체 생성 시 함께 처리됩니다.
체크해주셔서 감사합니다. 😄

parkeeseul commented on src/test/java/step3/domain/LottoTest.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    @Test
+    void invalid_로또_생성_숫자개수() {
+        assertThatThrownBy(() -> {
+            Lotto newLotto = new Lotto(Arrays.asList(1, 2, 3, 4, 5));
+        }).isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContaining("중복되지 않은 6개의 숫자를 입력해주세요.");
+    }
```
사용하지 않는 변수는 제거해 줘도 좋을 것 같아요 😄 
```suggestion
    void invalid_로또_생성_숫자개수() {
        assertThatThrownBy(() -> new Lotto(Arrays.asList(1, 2, 3, 4, 5)))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("중복되지 않은 6개의 숫자를 입력해주세요.");
    }
```

piopoi replied:
수정하였습니다. 😄

parkeeseul commented on src/main/java/step3/domain/LottoGame.java:
```diff
@@ -0,0 +1,40 @@
+        OutputView.printResult(lottoResult);
+    }
+
+    public static int buyCount(int money) {
+        return money / PRICE_LOTTO;
+    }
+
+    private List<Lotto> buy(int buyCount) {
+        System.out.printf("%d개를 구매했습니다.%n", buyCount);
+        List<Lotto> lottos = new ArrayList<>();
+        for (int i = 0; i < buyCount; i++) {
+            lottos.add(LottoFactory.createAutoLotto());
+        }
+        return lottos;
+    }
+
+    int investmentAmount(int buyCount) {
+        return buyCount * PRICE_LOTTO;
+    }
```
도메인 로직과 UI 로직이 혼재되어 있다고 생각되는데 어떻게 생각하시나요? 😄

piopoi replied:
UI 로직을 분리하여 UI를 전담하는 클래스로 모아두었습니다.
체크 감사합니다! 👍

parkeeseul commented on src/test/java/step3/domain/LottoNumberTest.java:
```diff
@@ -0,0 +1,24 @@
+import org.junit.jupiter.params.provider.ValueSource;
+
+public class LottoNumberTest {
+    @Test
+    void 로또번호_생성() {
+        Assertions.assertThat(LottoNumber.of(1)).isEqualTo(LottoNumber.of(1));
+    }
```
로또 번호 생성에 대한 테스트라기 보다는 동등성에 대한 테스트로 보여지는데 어떻게 생각하시나요? 😄

piopoi replied:
맞습니다. 캐시에서 가져오는 같은 객체를 사용하기 위한 테스트인데 메서드명을 잘못 썼네요... 😢 
수정하겠습니다.

parkeeseul replied:
@DisplayName 을 통해 테스트의 설명을 적어줘도 좋을 것 같아요 😄

piopoi replied:
토비님께서 영어에 자신 없으시면 한글로 생성하라고 하셔서 그 강의 이후부터는 한글로 메서드명을 생성하고 있습니다. 😄

parkeeseul commented on src/test/java/step3/domain/LottoNumberTest.java:
```diff
@@ -0,0 +1,24 @@
+        Assertions.assertThat(LottoNumber.of(1)).isEqualTo(LottoNumber.of(1));
+    }
+
+    @ParameterizedTest(name = "invalid_로또번호_생성")
+    @ValueSource(ints = {-1, 100, 0})
+    void invalid_로또번호_범위() {
```
테스트하고자 하는 범위의 경계값을 사용해 주면 어떨까요?
```suggestion
    @ParameterizedTest(name = "invalid_로또번호_생성")
    @ValueSource(ints = {0, 46})
    void invalid_로또번호_범위() {
```

piopoi replied:
실제 오류는 말씀하신 경계값에서 많이 겪는 것 같습니다.
테스트 값에도 더 고민해야 할 것 같습니다.
추가로 위 메서드에 문제가 있었네요..
수정하겠습니다. 😄

parkeeseul commented on src/test/java/step3/domain/LottoResultTest.java:
```diff
@@ -0,0 +1,33 @@
+        lottoResult = new LottoResult();
+    }
+
+    @Test
+    void 순위별_당첨횟수_확인() {
+        assertThat(lottoResult.rankingCount(Ranking.FIRST)).isEqualTo(0);
```
isZero() 를 사용해 볼 수도 있을 것 같네요 😄 
```suggestion
    @Test
    void 순위별_당첨횟수_확인() {
        assertThat(lottoResult.rankingCount(Ranking.FIRST)).isZero();
```

piopoi replied:
isZero를 계속 놓치네요 😢 
감사합니다!

parkeeseul reviewed (changes requested):
3단계 미션 잘 진행해 주셨네요! 👍 
몇 가지 코멘트 남겨두었으니 확인 부탁드립니다 😄 
또 한주가 시작되었네요!
즐거운 한주 되시길 바라요! 화이팅! 🔥

piopoi committed:
refactor(Lotto): 변수명 변경. LOTTO_SIZE -> SIZE

piopoi committed:
refactor(Lotto): 생성자 내 validate 메서드 위치 변경

piopoi committed:
refactor: 금액 값에 언더바로 천단위 구분 추가

piopoi committed:
refactor: 외부 접근이 필요한 상수의 접근제어자 변경 및 get 메서드 제거

piopoi committed:
refactor: domain 로직과 UI 로직 분리

piopoi committed:
refactor: WinnerLotto 분리

WinnerLotto
- 입력받은 당첨번호로 Lotto 객체를 생성하여 내부적으로 보유.

piopoi committed:
refactor: 불필요한 변수 제거

piopoi committed:
refactor: 객체 생성이 아닌 동등성 테스트가 목적인 메서드이기 때문에 메서드명 수정

piopoi committed:
refactor: 범위의 경계값을 테스트하도록 수정.

piopoi committed:
refactor: isEqualTo -> isZero 변경

parkeeseul commented on src/main/java/step3/domain/Lotto.java:
```diff
@@ -0,0 +1,64 @@
+                .count();
+    }
+
+    public List<LottoNumber> sortedLottoList() {
+        List<LottoNumber> lottoNumbers = new ArrayList<>(lotto);
+        Collections.sort(lottoNumbers);
+        return lottoNumbers;
+    }
```
메서드 네이밍에 타입이 담기게 되면 반환 타입이 변경될 경우
 메서드의 네이밍도 함께 변경 되어야 할 것 같은데 어떻게 생각하시나요?
또한 Lotto 를 생성할때 정렬을 해주면 어떨까 하는 생각도 드네요 😄

piopoi replied:
Lotto에 LottoNumber를 저장하기 위해 Set을 이용하기 때문에 Lotto 생성 시 정렬하려면 Lotto의 자료구조부터 변경해야 할 것 같습니다. 이름을 sortedLottoNumbers로 변경하겠습니다 😄

parkeeseul replied:
TreeSet 에 대해서도 한번 확인해 보셔도 좋을 것 같아요 😄

piopoi replied:
TreeSet에 대해 공부하겠습니다.
좋은 정보 감사합니다. 😄

parkeeseul commented on src/main/java/step3/domain/LottoResult.java:
```diff
@@ -0,0 +1,43 @@
+public class LottoResult {
+    private final EnumMap<Ranking, Integer> rankingMap;
+    private double winningAmount;
+    private double yield;
```
winningAmount , yield 는 상태값으로 관리될 필요가 있을까요? 😄

piopoi replied:
말씀하신대로 yield는 상태값으로 관리할 필요가 없어서 제거하였습니다.
다만, winningAmount는 누적 당첨금의 의미라서 로또게임결과(LottoResult)에서 관리하는게 맞다 싶어 남겨두었는데 수정하는게 좋을까요? 😄

parkeeseul replied:
winningAmount  는 calculateYield() 기능을 수행하기 위해 필요한 상태 값인 것 같은데요.
그렇다면 상태값으로 관리하지 않고 calculateYield()  에서 winningAmount 를 계산해서 기능을 수행해도 좋지 않을까 하는 
생각이 들었어요 😄

piopoi replied:
말씀하신 대로 winningAmount(당첨금총액)이 수익률 계산 외에서는 사용되지 않아서 상태값에서 제거하였습니다.
수익률 계산 시 당첨금 총액 계산 메서드를 호출하도록 수정하였습니다. 
조언 감사합니다 !! 😄

parkeeseul commented on src/main/java/step3/domain/WinnerLotto.java:
```diff
@@ -0,0 +1,32 @@
+
+import java.util.List;
+
+public class WinnerLotto {
+    private final Lotto lotto;
+
```
WinningLotto 와 Lotto는 어떤 차이가 있을까요? 😄

piopoi replied:
당첨로또 라는 정체성을 더 명확하게 표현하기 위해 생성하였습니다.
당첨로또 클래스 내의 자료구조를 생각하다보니 List<LottoNumber> 보다는 Lotto를 가지고 있게 되면 Lotto에서 수행하는 validate 로직들을 이용할 수 있어서 Lotto를 가지고 있는게 낫겟다 싶었습니다.
WinnerLotto를 제거하고 아래와 같이 변수명으로 당첨로또임을 표현하는 방향이 더 좋을까요? 😄 
`Lotto winnerLotto = new Lotto();`

parkeeseul replied:
그러셨군요!
당첨 로또라는 것을 표현해 주는 것은 좋지만 해당 클래스는 어떤 추가적인 책임이 있다기 보다는 
단순히 Lotto의 기능들을 바이패스 해주고 있어서 필요할까라는 의문이 들었습니다.

다음 단계에서는 WinnerLotto에 적절한 책임을 부여해 볼 수 있을 것 같네요

piopoi replied:
말씀하신대로 다음 단계에서 WinnerLotto에 리팩토링을 진행하려 합니다.
보너스 번호의 존재가 추가되면서 관련 로직을 WinnerLotto에 심어야 할 필요성을 느꼈습니다.
조언 감사합니다!! 😄

parkeeseul reviewed (changes requested):
피드백 잘 반영해 주셨네요! 👍 
몇 가지 추가 코멘트 남겨두었으니 확인 부탁드립니다 😄 
즐거운 오후 되세요! 화이팅 🔥

piopoi committed:
refactor: match 접근제어자 변경(default -> public)

piopoi committed:
refactor(Lotto): 메서드 명 변경(sortedLottoList -> sortedLotto)

piopoi committed:
refactor(LottoResult): 상태값 제거 (yield)

piopoi committed:
refactor(Lotto): 메서드명 변경(sortedLotto -> sortedLottoNumbers)

parkeeseul commented on src/main/java/step3/domain/WinnerLotto.java:
```diff
@@ -0,0 +1,32 @@
+        this.lotto = LottoFactory.createManualLotto(winnerNumbers);
+    }
+
+    public List<LottoNumber> lottoNumbers() {
+        return lotto.sortedLotto();
+    }
```
메서드 네이밍을 sortedLottoNumbers 으로 변경 후 사용하는 측에서 변경해주지 않으셨네요 😢

piopoi replied:
아이고.. 죄송합니다. 😢 
수정하겠습니다!!

parkeeseul reviewed (changes requested):
피드백 잘 반영해 주셨네요 👍 
몇 가지 코멘트 남겨두었으니 확인 부탁드립니다 😄 
즐거운 저녁 되세요 🔥

piopoi committed:
fix: 메서드명 변경 놓친 부분 수정(sortedLottoNumbers)

piopoi committed:
refactor(LottoResult): winningAmount 상태값 제거

1. LottoResult에서 winningAmount 상태값 제거
2. winningAmount(당첨금총액) 계산 메서드 추가
3. 관련 프로덕션/테스트 코드 수정

piopoi committed:
refactor(LottoResult): winningAmount 메서드 접근제어자 수정

piopoi reviewed (commented):
comment에 오타가 있네요.. 죄송합니다. 😢 
fa5ffba: [winning 메서드 mount접근제어자 수정] -> [winningAmount 메서드 접근제어자 수정]

rebase로 comment 수정하였습니다. 😄

parkeeseul reviewed (approved):
피드백 잘 반영해 주셨네요! 👍 
다음 단계 진행해 주세요! 
화이팅 입니다 🔥
